src/hash/blake/mod.rs

```diff
@@ -0,0 +1,226 @@
+        hasher.update(&seed.0);
+        hasher.update(&value.to_le_bytes());
+        Blake3Digest(hasher.finalize().into())
+    }
+}
+
+impl ElementHasher for Blake3_256 {
+    type BaseField = Felt;
+
+    fn hash_elements<E>(elements: &[E]) -> Self::Digest
+    where
+        E: FieldElement<BaseField = Self::BaseField>,
+    {
+        Blake3Digest(shrink_elements_bytes(elements))
+    }
+}
+
+// BLAKE3 192-BIT OUTPUT
+// ================================================================================================
+
+/// 192-bit output blake3 hasher.
+pub struct Blake3_192;
+
+impl HashFn for Blake3_192 {
+    type Digest = Blake3Digest<24>;
+
+    fn hash(bytes: &[u8]) -> Self::Digest {
+        Blake3Digest(*shrink_bytes(&blake3::hash(bytes).into()))
+    }
+
+    fn merge(values: &[Self::Digest; 2]) -> Self::Digest {
+        Self::hash(prepare_merge(values))
+    }
+
+    fn merge_with_int(seed: Self::Digest, value: u64) -> Self::Digest {
+        let mut hasher = blake3::Hasher::new();
+        hasher.update(&seed.0);
+        hasher.update(&value.to_le_bytes());
+        Blake3Digest(*shrink_bytes(&hasher.finalize().into()))
+    }
+}
+
+impl ElementHasher for Blake3_192 {
+    type BaseField = Felt;
+
+    fn hash_elements<E>(elements: &[E]) -> Self::Digest
+    where
+        E: FieldElement<BaseField = Self::BaseField>,
+    {
+        Blake3Digest(shrink_elements_bytes(elements))
+    }
+}
+
+// BLAKE3 160-BIT OUTPUT
+// ================================================================================================
+
+/// 160-bit output blake3 hasher.
+pub struct Blake3_160;
+
+impl HashFn for Blake3_160 {
+    type Digest = Blake3Digest<20>;
+
+    fn hash(bytes: &[u8]) -> Self::Digest {
+        Blake3Digest(*shrink_bytes(&blake3::hash(bytes).into()))
+    }
+
+    fn merge(values: &[Self::Digest; 2]) -> Self::Digest {
+        Self::hash(prepare_merge(values))
+    }
+
+    fn merge_with_int(seed: Self::Digest, value: u64) -> Self::Digest {
+        let mut hasher = blake3::Hasher::new();
+        hasher.update(&seed.0);
+        hasher.update(&value.to_le_bytes());
+        Blake3Digest(*shrink_bytes(&hasher.finalize().into()))
+    }
+}
+
+impl ElementHasher for Blake3_160 {
+    type BaseField = Felt;
+
+    fn hash_elements<E>(elements: &[E]) -> Self::Digest
+    where
+        E: FieldElement<BaseField = Self::BaseField>,
+    {
+        Blake3Digest(shrink_elements_bytes(elements))
+    }
+}
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+/// Zero-copy ref shrink to array
+fn shrink_bytes<const M: usize, const N: usize>(bytes: &[u8; M]) -> &[u8; N] {
+    // compile-time assertion
+    assert!(M >= N);
+    // safety: bytes len is asserted
+    unsafe { transmute(bytes) }
+}
+
+/// Hash the elements into bytes and shrinz the output
+fn shrink_elements_bytes<const N: usize, E>(elements: &[E]) -> [u8; N]
```

Comment:
nit: I found the name a bit confusing since the main operation is hashing rather than shrinking. Maybe this should be `hash_elements()`?

Reply:
Fixed!